lib/fastlane/plugin/wpmreleasetoolkit/models/firebase_test_runner.rb

```diff
@@ -0,0 +1,88 @@
+require 'json'
+require 'uri'
+require 'fileutils'
+
+module Fastlane
+  class FirebaseTestRunner
+    VALID_TEST_TYPES = %w[instrumentation robo].freeze
+
+    def initialize(key_file:, automatic_login: true)
+      raise "Unable to find key file: #{key_file}" unless File.file? key_file
+
+      @key_file = key_file
+      authenticate if automatic_login
+    end
+
+    def authenticate
+      Action.sh(
+        'gcloud', 'auth', 'activate-service-account',
+        '--key-file', @key_file
+      )
+    end
+
+    def run_tests(apk_path:, test_apk_path:, device:, type: 'instrumentation')
+      raise "Unable to find apk: #{apk_path}" unless File.file? apk_path
+      raise "Unable to find apk: #{test_apk_path}" unless File.file? test_apk_path
+      raise "Invalid Type: #{type}" unless VALID_TEST_TYPES.include? type
+
+      command = [
+        'gcloud', 'firebase', 'test', 'android', 'run',
+        '--type', Shellwords.escape(type),
+        '--app', Shellwords.escape(apk_path),
+        '--test', Shellwords.escape(test_apk_path),
+        '--device', Shellwords.escape(device.to_s),
+        '--verbosity', 'info',
+      ].join(' ')
+
+      log_file_path = Fastlane::Actions.lane_context[:FIREBASE_TEST_LOG_FILE_PATH]
+
+      UI.message "Streaming log output to #{log_file_path}"
+      Action.sh("#{command} 2>&1 | tee #{log_file_path}")
+
+      # Make the file object available to other tasks
+      result = FirebaseTestLabResult.new(log_file_path: log_file_path)
+      Fastlane::Actions.lane_context[:FIREBASE_TEST_LOG_FILE] = result
+
+      result
+    end
+
+    def download_result_files(result:, destination:, project_id:, key_file_path:)
+      raise unless result.is_a? Fastlane::FirebaseTestLabResult
+
+      paths = result.raw_results_paths
+      raise "Log File doesn't contain a raw results URL" if paths.nil?
+
+      FileUtils.mkdir_p(destination) unless File.directory? destination
+
+      require 'google/cloud/storage'
```

Comment:
💄  Move this `require` to the top of the file

Reply:
Is lazy-loading mostly not a thing in Ruby?

Addressed in 7ac5002af00ea12db03283f555e5cd23b4663286

Reply:
Yeah it has value to only require in the method body so that the lib/gem is only loaded the first time the code is executed (and thus never loaded if never used).

But I think in practice it's seldom used and the conventions prefers grouping all `require` at top to make it easier to see all the dependencies in one place?

Though I haven't found anything about such a rule in the https://rubystyle.guide so maybe it's more of a _de facto_ common usage more than anything?

Anyway, I think having `require` statements grouped at top makes it easier to reason about your overall dependencies, and moving it inside a method call for lazy loading could be useful for performance… but is likely premature optimization at this point?